src/backend-function.lisp

```diff
@@ -0,0 +1,221 @@
+;;;; matrix arithmetic routines. For instance, we might want both Lisp
+;;;; and C implementations of the same function, and we might want the
+;;;; user to be able to select such functions at run-time.
+;;;;
+;;;; Terminology:
+;;;;
+;;;;     BACKEND: A name of a backend, usually a keyword. Make a new
+;;;;     one with DEFINE-BACKEND.
+;;;;
+;;;;     BACKEND FUNCTION: A function that "wraps"
+;;;;     implementations. This is sort of like a Common Lisp generic
+;;;;     function. Make a new one with DEFINE-BACKEND-FUNCTION.
+;;;;
+;;;;     BACKEND IMPLEMENTATION: A function that actually contains the
+;;;;     code for a specific function's backend. This is sort of like
+;;;;     a Common Lisp method. Make a new one with
+;;;;     DEFINE-BACKEND-IMPLEMENTATION.
+
+;;; first some utilities
+
+(deftype function-designator ()
+  '(or symbol function))
+
+(defun normalize-optional/keyword-argument (arg)
+  (destructuring-bind (arg init supplied) (alexandria:ensure-list arg)
+    (list arg init (if (null supplied)
+                       (gensym "SUPPLIEDP")
+                       supplied))))
+
+(defun normalize-lambda-list (lambda-list)
+  (multiple-value-bind (required-args
+                        optional-args                 ; List of (NAME INIT SUPPLIEDP)
+                        rest-arg
+                        kw-args                  ; List of ((KW NAME) INIT SUPPLIEDP)
+                        allow-other-keys
+                        aux-args)
+      (alexandria:parse-ordinary-lambda-list lambda-list)
+    (let ((normalized required-args))
+      (when rest-arg
+        (setf normalized (append normalized '(&rest) (list rest-arg))))
+      (when optional-args
+        (setf normalized
+              (append normalized '(&optional)
+                      (mapcar #'normalize-optional/keyword-argument optional-args))))
+      (when kw-args
+        (setf normalized
+              (append normalized '(&key)
+                      (mapcar #'normalize-optional/keyword-argument kw-args))))
+      (when allow-other-keys
+        (setf normalized (append normalized '(&allow-other-keys))))
+      (when aux-args
+        (setf normalized (append normalized '(&aux) aux-args)))
+      normalized)))
+
+;;; now the meat
+
+(define-condition no-applicable-implementation (simple-error)
+  ()
+  (:documentation "The error to signal when there's no applicable implementation, whether from a backend function or a CLOS generic function."))
+
+(defun no-applicable-implementation (name)
+  (error 'no-applicable-implementation
+         :format-control (format nil "~S is not implemented in any of ~
+                                      the current active backends: ~
+                                      ~~{~~A~~^, ~~}."
+                                 name)
+         :format-arguments (list (active-backends))))
+
+(defmacro define-compatible-no-applicable-method-behavior (&rest generic-function-names)
```

Comment:
I know this is shown in one of the Markdown writeups, but can we have a docstring on this (or some mention in the comments at the top)? Even just a remark about what "compatible" means.

Reply:
done